src/style/style.js

```diff
@@ -568,7 +568,12 @@ class Style extends Evented {
 
         layer.setEventedParent(this, {layer: {id: id}});
 
-        const index = before ? this._order.indexOf(before) : this._order.length;
+
+        const beforeIndex = before ? this._order.indexOf(before) : -1;
+        const index = before && beforeIndex !== -1 ? beforeIndex : this._order.length;
+        if (before && beforeIndex === -1) this.fire('error', {
```

Comment:
Instead of adding to the top of the stack, return early without making changes in the case of an error.